webrender/src/tiling.rs

```diff
@@ -143,37 +124,74 @@ impl BatchList {
         match key.kind {
             AlphaBatchKind::Composite { .. } => {}
             _ => {
-                'outer: for (batch_index, batch) in batches.iter()
-                                                           .enumerate()
-                                                           .rev()
-                                                           .take(10) {
+                'outer: for (batch_index, batch) in self.batches
+                                                        .iter()
+                                                        .enumerate()
+                                                        .rev()
+                                                        .take(10) {
                     if batch.key.is_compatible_with(key) {
                         selected_batch_index = Some(batch_index);
                         break;
                     }
 
                     // check for intersections
-                    if check_intersections {
-                        for item_rect in &batch.item_rects {
-                            if item_rect.intersects(item_bounding_rect) {
-                                break 'outer;
-                            }
+                    for item_rect in &batch.item_rects {
+                        if item_rect.intersects(item_bounding_rect) {
+                            break 'outer;
                         }
                     }
                 }
             }
         }
 
         if selected_batch_index.is_none() {
-            let new_batch = PrimitiveBatch::new(key.clone());
-            selected_batch_index = Some(batches.len());
-            batches.push(new_batch);
+            let new_batch = AlphaPrimitiveBatch::new(key.clone());
+            selected_batch_index = Some(self.batches.len());
+            self.batches.push(new_batch);
         }
 
-        let batch = &mut batches[selected_batch_index.unwrap()];
+        let batch = &mut self.batches[selected_batch_index.unwrap()];
         batch.item_rects.push(*item_bounding_rect);
 
-        batch
+        &mut batch.instances
+    }
+}
+
+pub struct OpaqueBatchList {
```

Comment:
Although the code is mostly duplicated right now, I'm planning to experiment with changing the batching strategies for opaque and alpha - I think they'll end up being quite different.

In particular, while the alpha batcher would stay quite similar, the opaque batcher would ensure very large primitives get rendered first (even if a batch break is involved), while smaller primitives could get added to a per-batch-type list, without having to search for an appropriate batch.

So I think it makes sense to leave the code duplicated for now - does that seem reasonable?